exchanges/kucoin/kucoin_websocket.go

```diff
@@ -0,0 +1,1645 @@
+			marketMatchChannel,
+			markPriceIndicatorChannel:
+			symbols, okay := subscriptions[x].Params["symbols"].(string)
+			if !okay {
+				if subscriptions[x].Currency.IsEmpty() {
+					return nil, errors.New("symbols not passed")
+				}
+				symbols = subscriptions[x].Currency.String()
+			}
+			payloads[x] = WsSubscriptionInput{
+				ID:       strconv.FormatInt(ku.Websocket.Conn.GenerateMessageID(false), 10),
+				Type:     operation,
+				Topic:    fmt.Sprintf(subscriptions[x].Channel, symbols),
+				Response: true,
+			}
+		case marketAllTickersChannel,
+			privateChannel,
+			accountBalanceChannel,
+			marginPositionChannel,
+			spotMarketAdvancedChannel,
+			futuresTradeOrderChannel,
+			futuresStopOrdersLifecycleEventChannel,
+			futuresAccountBalanceEventChannel, futuresSystemAnnouncementChannel:
+			input := WsSubscriptionInput{
+				ID:       strconv.FormatInt(ku.Websocket.Conn.GenerateMessageID(false), 10),
+				Type:     operation,
+				Topic:    subscriptions[x].Channel,
+				Response: true,
+			}
+			switch subscriptions[x].Channel {
+			case futuresTradeOrderChannel,
+				futuresStopOrdersLifecycleEventChannel,
+				futuresAccountBalanceEventChannel:
+				input.PrivateChannel = true
+			}
+			payloads[x] = input
+		case marketTickerSnapshotChannel,
+			futuresPositionChangeEventChannel,
+			futuresTradeOrdersBySymbolChannel,
+			futuresTrasactionStatisticsTimerEventChannel,
+			futuresContractMarketDataChannel,
+			futuresOrderbookLevel2Depth50Channel,
+			futuresOrderbookLevel2Depth5Channel,
+			futuresExecutionDataChannel,
+			futuresOrderbookLevel2Channel,
+			futuresTickerChannel,
+			futuresTickerV2Channel: // Symbols
+			payloads[x] = WsSubscriptionInput{
+				ID:       strconv.FormatInt(ku.Websocket.Conn.GenerateMessageID(false), 10),
+				Type:     operation,
+				Topic:    fmt.Sprintf(subscriptions[x].Channel, subscriptions[x].Currency.String()),
+				Response: true,
+			}
+			switch subscriptions[x].Channel {
+			case futuresPositionChangeEventChannel,
+				futuresTradeOrdersBySymbolChannel:
+				payloads[x].PrivateChannel = true
+			}
+		case marketTickerSnapshotForCurrencyChannel,
+			marginLoanChannel:
+			if subscriptions[x].Channel == marketTickerSnapshotForCurrencyChannel {
+				subscriptions[x].Channel += "%s"
+			}
+			payloads[x] = WsSubscriptionInput{
+				ID:       strconv.FormatInt(ku.Websocket.Conn.GenerateMessageID(false), 10),
+				Type:     operation,
+				Topic:    fmt.Sprintf(subscriptions[x].Channel, subscriptions[x].Currency.Base.Upper().String()),
+				Response: true,
+			}
+		case marketCandlesChannel:
+			interval, err := ku.intervalToString(subscriptions[x].Params["interval"].(kline.Interval))
+			if err != nil {
+				return nil, err
+			}
+			payloads[x] = WsSubscriptionInput{
+				ID:       strconv.FormatInt(ku.Websocket.Conn.GenerateMessageID(false), 10),
+				Type:     operation,
+				Topic:    fmt.Sprintf(subscriptions[x].Channel, subscriptions[x].Currency.Base.Upper().String(), interval),
```

Comment:
Why did you decide to make it send just the currency base when the documentation does not suggest doing so?
KuCoin sends back ack's for subscriptions, but it appears it does no validation, so you're sending `"/market/candles:BTC_1hour"` instead of `"/market/candles:BTC-USDT_1hour"`.

Even if you were supposed to only send the base, you're not checking if the base has already been sent. So you could send many duplicate BTC subscriptions as a result of this. You should have already tested this and seen that you get a response before putting this up for review

Reply:
Ohh, man. I have no idea why I didn't see that. I thought the acknowledge message was a success message and was not returning the candle data because of the currencies. Anyways, I will make sure to not create such kind of mistakes.

I have tested and amended the channel with Full instrument and it's working now. except for some instruments which haven't candlestick data.